wgpu-core/src/command/bind.rs

```diff
@@ -142,14 +144,21 @@ mod compat {
 
                         let mut errors = Vec::new();
 
-                        let mut expected_bgl_entries = expected_bgl.entries.iter();
-                        let mut assigned_bgl_entries = assigned_bgl.entries.iter();
+                        let mut expected_bgl_entries = expected_bgl.entries.values();
+                        let mut assigned_bgl_entries = assigned_bgl.entries.values();
                         let zipped = crate::utils::ZipWithProperAdvance::new(
                             &mut expected_bgl_entries,
                             &mut assigned_bgl_entries,
                         );
 
-                        for ((&binding, expected_entry), (_, assigned_entry)) in zipped {
+                        for (expected_entry, assigned_entry) in zipped {
+                            if assigned_entry.binding != expected_entry.binding {
+                                errors.push(EntryError::Binding {
+                                    expected: expected_entry.binding,
+                                    assigned: assigned_entry.binding,
+                                });
+                            }
+                            let binding = expected_entry.binding;
```

Comment:
@teoxoy this is somehow weird, as each entry has different binding number. I think `if assigned_entry.binding != expected_entry.binding {` we should just continue and not print other errors, what do you think?

Reply:
Yeah, we shouldn't zip those at all actually.

Reply:
```rust
for expected_entry in expected_bgl.entries.values() {
    if let Some(assigned_entry) =
        assigned_bgl.entries.get(expected_entry.binding)
    {
        if assigned_entry.visibility != expected_entry.visibility {
            errors.push(EntryError::Visibility {
                binding: expected_entry.binding,
                expected: expected_entry.visibility,
                assigned: assigned_entry.visibility,
            });
        }
        if assigned_entry.ty != expected_entry.ty {
            errors.push(EntryError::Type {
                binding: expected_entry.binding,
                expected: expected_entry.ty,
                assigned: assigned_entry.ty,
            });
        }
        if assigned_entry.count != expected_entry.count {
            errors.push(EntryError::Count {
                binding: expected_entry.binding,
                expected: expected_entry.count,
                assigned: assigned_entry.count,
            });
        }
    } else {
        errors.push(EntryError::ExtraExpected {
            binding: expected_entry.binding,
        });
    }
}

for assigned_entry in assigned_bgl.entries.values() {
    if !expected_bgl.entries.contains_key(assigned_entry.binding) {
        errors.push(EntryError::ExtraAssigned {
            binding: assigned_entry.binding,
        });
    }
}
```

I got this working locally.

Reply:
I just checked in spec and I think we can, see https://gpuweb.github.io/gpuweb/#bind-group-compatibility (either it must exists in both or it must be missing on both)

Reply:
> I got this working locally.

So basically https://github.com/gfx-rs/wgpu/commit/4a19ac279c4f81aacedb1d215c884c10fe115275#diff-d89ef4ca07007b9044c6ca29c1632570d5e437de81fcaafde8807ebd894d6e9bL81 but with new error types.

Reply:
The issue is that if we zip the iterators we'll end up with errors that are not accurate in situations like:
expected_bgl with bindings: 0, 1, 2, 3, 4
assigned_bgl with bindings: 0, 2, 3, 4
Binding 1 will be compared with 2, 2 with 3 and 3 with 4 and we'll say 4 is missing.

If we don't zip, then we will only say that binding 1 is missing.